File: packages/core/android/src/main/java/com/posedetection/export/ExportOptions.kt
package com.posedetection.export

import android.content.Context
import android.net.Uri
import com.posedetection.view.OverlayConfig
import com.posedetection.view.parseOverlay
import java.io.File

/** What `exportPose` was asked for. Defaults from `guides/export.md`. */
internal class ExportOptions(
    /**
     * The same config the camera's `overlay` prop takes, so a painted file and a live preview are
     * configured with one vocabulary rather than two.
     */
    val overlay: OverlayConfig,
    val drawOverlay: Boolean,
    val maxPoses: Int,
    /**
     * Detection samples a second. Between samples the last pose is held, exactly as the live
     * overlay holds one between inferences.
     */
    val sampleFps: Int,
    /** Long edge of the output, or 0 for the source's own size. */
    val maxSize: Int,
    val directory: File,
    val fileName: String,
    /** Still images only. */
    val quality: Int,
) {
    companion object {
        private const val DEFAULT_SAMPLE_FPS = 10
        private const val MIN_MAX_SIZE = 120

        fun parse(
            context: Context,
            raw: Map<*, *>?,
            sourceName: String,
        ): ExportOptions {
            val overlayRaw = raw?.get("overlay")
            return ExportOptions(
                overlay = (overlayRaw as? Map<*, *>)?.let { parseOverlay(it) } ?: OverlayConfig(),
                drawOverlay = overlayRaw as? Boolean ?: true,
                maxPoses = count(raw?.get("maxPoses"), 1, 5),
                sampleFps = count(raw?.get("fps"), DEFAULT_SAMPLE_FPS, 60),
                maxSize = maxSize(raw?.get("maxSize")),
                directory = directory(context, raw?.get("directory") as? String),
                fileName = fileName(raw?.get("fileName") as? String, sourceName),
                quality =
                    ((raw?.get("quality") as? Number)?.toFloat() ?: 0.9f)
                        .coerceIn(0.1f, 1f)
                        .let { (it * 100).toInt() },
            )
        }

        /**
         * Where the file lands, created if it is not there yet.
         *
         * The default is the app's cache directory: an export is derived data, and a package that
         * wrote into the app's files directory by default would leave behind copies the user never
         * asked for and never sees. Apps that want it kept pass a directory of their own, which is
         * also how the file ends up somewhere they can upload or move it from.
         */
        private fun directory(
            context: Context,
            raw: String?,
        ): File {
            val base =
                when (raw) {
                    null, "cache" -> {
                        context.cacheDir
                    }

                    "documents" -> {
                        context.filesDir
                    }

                    else -> {
                        val parsed = Uri.parse(raw)
                        File(if (parsed.scheme == "file") parsed.path ?: raw else raw)
                    }
                }
            base.mkdirs()
            return base
        }

        /**
         * Sanitized rather than trusted: this reaches the filesystem, and a name with a slash in it
         * would write outside the directory the caller chose.
         */
        private fun fileName(
            raw: String?,
            sourceName: String,
        ): String {
            val candidate = raw?.takeIf { it.isNotEmpty() } ?: "$sourceName-pose"
            val cleaned =
                candidate
                    .filter { it.isLetterOrDigit() || it == '-' || it == '_' || it == ' ' || it == '.' }
                    .trim()
            return cleaned.ifEmpty { "pose-export" }
        }

        private fun maxSize(value: Any?): Int {
            val size = (value as? Number)?.toInt() ?: return ExportCanvas.DEFAULT_MAX_SIZE
            return if (size <= 0) 0 else size.coerceAtLeast(MIN_MAX_SIZE)
        }

        private fun count(
            value: Any?,
            fallback: Int,
            limit: Int,
        ): Int = (value as? Number)?.toInt()?.coerceIn(1, limit) ?: fallback
    }
}

/** What came back, for the JavaScript side to turn into an `ExportResult`. */
internal class ExportSummary(
    val file: File,
    val width: Int,
    val height: Int,
    val durationMs: Int,
    val frameCount: Int,
    val posesFound: Int,
) {
    fun payload(): Map<String, Any> =
        mapOf(
            "uri" to Uri.fromFile(file).toString(),
            "width" to width,
            "height" to height,
            "durationMs" to durationMs,
            "frameCount" to frameCount,
            "posesFound" to posesFound,
        )
}

internal class ExportError(
    message: String,
) : Exception(message)

/** Thrown when the caller cancelled. Distinct from a failure, because it is not one. */
internal class ExportCancelled : Exception("the export was cancelled")

File: packages/core/ios/export/ExportOptions.swift
import Foundation
import UIKit

/// What `exportPose` was asked for. Defaults from guides/export.md.
struct ExportOptions {
  /// The same config the camera's `overlay` prop takes, so a painted file and a live preview are
  /// configured with one vocabulary rather than two.
  let overlay: OverlayConfig
  let drawOverlay: Bool
  let maxPoses: Int
  /**
   How sure the model has to be before it calls something a body.

   Left out, it follows `maxPoses`, because the two are one decision: 0.5 for a single subject,
   which is MediaPipe's own, and 0.3 above that, which is where a second person actually appears
   rather than the first person twice. A number overrides it, and what the right number is for a
   given piece of footage is the caller's to know.
   */
  let minConfidence: Float
  /// Detection samples a second. Between samples the last pose is held, exactly as the live
  /// overlay holds one between inferences.
  let sampleFps: Int
  /// Long edge of the output, or 0 for the source's own size.
  let maxSize: Int
  let directory: URL
  let fileName: String
  /// Still images only.
  let quality: CGFloat

  static let defaultMaxSize = 1920
  static let defaultSampleFps = 10

  static func parse(_ raw: [String: Any]?, sourceName: String) throws -> ExportOptions {
    let maxPoses = clampedCount(raw?["maxPoses"], fallback: 1, limit: 5)
    var drawOverlay = true
    var overlay = OverlayConfig()
    if let value = JS.bool(raw?["overlay"]) {
      drawOverlay = value
    } else if let map = JS.dictionary(raw?["overlay"]) {
      overlay = parseOverlay(map)
    }

    return ExportOptions(
      overlay: overlay,
      drawOverlay: drawOverlay,
      maxPoses: maxPoses,
      minConfidence: minConfidence(raw?["minConfidence"], maxPoses: maxPoses),
      sampleFps: clampedCount(raw?["fps"], fallback: defaultSampleFps, limit: 60),
      maxSize: maxSize(raw?["maxSize"]),
      directory: try directory(JS.string(raw?["directory"])),
      fileName: fileName(JS.string(raw?["fileName"]), sourceName: sourceName),
      quality: CGFloat(clamped(JS.number(raw?["quality"]) ?? 0.9, 0.1, 1, 0.9))
    )
  }

  private static func minConfidence(_ raw: Any?, maxPoses: Int) -> Float {
    let auto = Double(PoseDetector.stillConfidence(forMaxPoses: maxPoses))
    return Float(clamped(JS.number(raw) ?? auto, 0.1, 0.9, auto))
  }

  /**
   Where the file lands, created if it is not there yet.

   The default is the app's caches directory: an export is derived data, and a package that wrote
   into Documents by default would put files the user never asked for into their iCloud backup.
   Apps that want it kept pass a directory of their own, which is also how the file ends up
   somewhere they can upload or move it from.
   */
  private static func directory(_ raw: String?) throws -> URL {
    let base: URL
    switch raw {
    case nil, "cache":
      base = try FileManager.default.url(
        for: .cachesDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
    case "documents":
      base = try FileManager.default.url(
        for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
    case let value?:
      guard let url = URL(string: value), url.isFileURL else {
        guard value.hasPrefix("/") else {
          throw ExportError("directory must be 'cache', 'documents' or a file:// URI, got \(value)")
        }
        base = URL(fileURLWithPath: value)
        try FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        return base
      }
      base = url
    }
    try FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
    return base
  }

  /// Sanitized rather than trusted: this reaches the filesystem, and a name with a slash in it
  /// would write outside the directory the caller chose.
  private static func fileName(_ raw: String?, sourceName: String) -> String {
    let candidate = raw.flatMap { $0.isEmpty ? nil : $0 } ?? "\(sourceName)-pose"
    let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "-_ ."))
    let cleaned = candidate.unicodeScalars.filter { allowed.contains($0) }.map(String.init).joined()
    let trimmed = cleaned.trimmingCharacters(in: .whitespaces)
    return trimmed.isEmpty ? "pose-export" : trimmed
  }

  private static func maxSize(_ value: Any?) -> Int {
    guard let number = JS.number(value) else { return defaultMaxSize }
    let size = Int(number)
    return size <= 0 ? 0 : max(120, size)
  }

  private static func clampedCount(_ value: Any?, fallback: Int, limit: Int) -> Int {
    guard let number = JS.number(value) else { return fallback }
    return min(max(1, Int(number)), limit)
  }
}

struct ExportError: LocalizedError {
  let message: String

  init(_ message: String) {
    self.message = message
  }

  var errorDescription: String? {
    return message
  }
}

/// What came back, for the JavaScript side to turn into an `ExportResult`.
struct ExportSummary {
  let url: URL
  let width: Int
  let height: Int
  let durationMs: Int
  let frameCount: Int
  let posesFound: Int

  var payload: [String: Any] {
    return [
      "uri": url.absoluteString,
      "width": width,
      "height": height,
      "durationMs": durationMs,
      "frameCount": frameCount,
      "posesFound": posesFound
    ]
  }
}
